src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+        // Note that the threading control should know about its schedulers for cancellation/exception/priority propagation,
+        // see e.g. task_group_context::cancel_group_execution()
+        while (my_public_ref_count.load(std::memory_order_acquire) == 1 && my_ref_count.load(std::memory_order_acquire) > 1) {
+            yield();
+        }
+        lock.acquire(g_threading_control_mutex);
+    }
+}
+
+bool threading_control::release(bool is_public, bool blocking_terminate) {
+    bool do_release = false;
+    {
+        global_mutex_type::scoped_lock lock(g_threading_control_mutex);
+        if (blocking_terminate) {
+            __TBB_ASSERT(is_public, "Only an object with a public reference can request the blocking terminate");
+            wait_last_reference(lock);
+        }
+        do_release = remove_ref(is_public);
+    }
+
+    if (do_release) {
+        __TBB_ASSERT(!my_public_ref_count.load(std::memory_order_relaxed), "No public references remain if we remove the threading control.");
+        // inform RML that blocking termination is required
+        my_pimpl->release(blocking_terminate);
+        return blocking_terminate;
+    }
+    return false;
+}
+
+threading_control::threading_control(unsigned public_ref, unsigned ref) : my_public_ref_count(public_ref), my_ref_count(ref)
+{}
+
+threading_control* threading_control::register_public_reference() {
+    threading_control* control{nullptr};
+    global_mutex_type::scoped_lock lock(g_threading_control_mutex);
+    control = get_threading_control(/*public = */ true);
+    if (!control) {
+        // We are going to create threading_control_impl, we should acquire mutexes in right order
+        lock.release();
+        control = create_threading_control();
+    }
+
+    return control;
+}
+
+bool threading_control::unregister_public_reference(bool blocking_terminate) {
+    __TBB_ASSERT(g_threading_control, "Threading control should exist until last public reference");
+    __TBB_ASSERT(g_threading_control->my_public_ref_count.load(std::memory_order_relaxed), nullptr);
+    return g_threading_control->release(/*public = */ true, /*blocking_terminate = */ blocking_terminate);
+}
+
+threading_control::threading_control_client threading_control::create_client(arena& a) {
+    {
+        global_mutex_type::scoped_lock lock(g_threading_control_mutex);
+        add_ref(/*public = */ false);
+    }
+
+    return my_pimpl->create_client(a);
+}
+
+void threading_control::publish_client(threading_control::threading_control_client client) {
+    return my_pimpl->publish_client(client);
+}
+
+bool threading_control::check_client_priority(threading_control::threading_control_client client) {
+    return my_pimpl->check_client_priority(client);
+}
+
+threading_control::client_deleter threading_control::prepare_client_destruction(threading_control::threading_control_client client) {
+    return my_pimpl->prepare_client_destruction(client);
+}
```

Comment:
Again, it seems there will be more value if we implement inheritance rather than pointer to implementation. Am I wrong? For example, we will just not have these methods implemented.

Reply:
Inheritance is really powerful tool and should be used accurately (should follow generalization principles e.g. [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle)). Since you propose to use it only to remove code duplication I prefer not to do so.